control/main.py

```diff
@@ -0,0 +1,194 @@
+"""Main program for autonomous drone flight and sensor data logging.
+
+Uses the ttyO4 connection (UART4) on the beaglebone to connect to a Pixhawk flight controller.
+Uses the ttyO1 connection (UART1) to connect to a xBee wireless communication module.
+Uses tcp://localhost:5555 to connect to zmq server for reading i2c sensor data.
+Installing the appropriate main.service file allows for this to start on boot.
+"""
+
+import logging
+import sys
+import time
+import dronekit
+from control.communication import Communication
+from control.controller import Controller
+from control.i2cdataclient import I2cDataClient
+from control.gps import get_location_offset, get_distance, get_relative_from_location
+from control.helper import location_global_relative_to_gps_reading, gps_reading_to_location_global
+
+MAX_RADIUS = 50
+MAX_ALTITUDE = 20
+MIN_ALTITUDE = 3
+
+
+def create_waypoints(logger, com, start_location, waypoints):
+    start_gps = location_global_relative_to_gps_reading(start_location)
+    location_points = []
+    for point in waypoints:
+        x = point[u'x']
+        y = point[u'y']
+        z = point[u'z']
+        gps_waypoint = get_location_offset(start_gps, y, x)
+        gps_waypoint.altitude = z
+        if get_distance(start_gps, gps_waypoint) > MAX_RADIUS:
+            logger.critical("Waypoint is {}".format(point))
+            logger.critical("Waypoint exceeds max allowed radius of {}m".format(MAX_RADIUS))
+            com.send("Waypoint is {}".format(point))
+            com.send("Waypoint exceeds max allowed radius of {}m".format(MAX_RADIUS))
+            return None
+        elif gps_waypoint.altitude > MAX_ALTITUDE:
+            logger.critical("Waypoint is {}".format(point))
+            logger.critical("Waypoint exceeds max allowed altitude of {}m".format(MAX_ALTITUDE))
+            com.send("Waypoint is {}".format(point))
+            com.send("Waypoint exceeds max allowed altitude of {}m".format(MAX_ALTITUDE))
+            return None
+        elif gps_waypoint.altitude < MIN_ALTITUDE:
+            logger.critical("Waypoint is {}".format(point))
+            logger.critical("Waypoint under min allowed altitude of {}m".format(MIN_ALTITUDE))
+            com.send("Waypoint is {}".format(point))
+            com.send("Waypoint under min allowed altitude of {}m".format(MIN_ALTITUDE))
+            return None
+        location_points.append(gps_reading_to_location_global(gps_waypoint))
+    return location_points
+
+
+def package_data(home, location, data_client):
```

Comment:
Another docstring, it's hard to tell if the function takes a LocationGlobalRelative without one. Same kind of deal goes for the previous function.